File: src/main/java/com/vahoss/java_solutions/Solution00125.java
package com.vahoss.java_solutions;

public class Solution00125 {
    public boolean isPalindrome(String s) {
        // Two pointer approach:
        var beg = 0;
        var end = s.length() - 1;

        while (beg < end) {
            while (beg < end && !Character.isLetterOrDigit(s.charAt(beg))) beg++;
            while (beg < end && !Character.isLetterOrDigit(s.charAt(end))) end--;
            if (Character.toLowerCase(s.charAt(beg)) != Character.toLowerCase(s.charAt(end))) return false;
            beg++;
            end--;
        }

        return true;
    }

    public boolean isPalindrome2(String s) {
        // 1. prepare the s
        var sb = new StringBuilder(s.toLowerCase().replaceAll("[^a-zA-Z0-9]", ""));
        return (sb.toString().contentEquals(sb.reverse()));
    }

    public static void main(String[] args) {
        var solution = new Solution00125();

        TestCase[] testCases = {
                new TestCase("racecar", true),
                new TestCase("race a car", false),
                new TestCase("A man, a plan, a canal, Panama", true),
                new TestCase("hello", false),
                new TestCase("Madam, I'm Adam", true),
                new TestCase("12321", true),
                new TestCase("OpenAI", false),
                new TestCase("Was it a car or a cat I saw?", true),
                new TestCase("Java", false),
                new TestCase("", true),
                new TestCase("a", true),
                new TestCase("Able was I ere I saw Elba", true)
        };

        int passedTests = 0;
        int totalTests = testCases.length;

        for (TestCase test : testCases) {
            boolean result = solution.isPalindrome(test.input);
            boolean passed = (result == test.expected);
            passedTests += passed ? 1 : 0;

            System.out.printf("Test: \"%s\"%n", test.input);
            System.out.printf("Expected: %b, Got: %b, Result: %s%n",
                    test.expected, result, passed ? "PASS" : "FAIL");
            System.out.println();
        }

        System.out.printf("Passed %d out of %d tests%n", passedTests, totalTests);

    }

    static class TestCase {
        String input;
        boolean expected;

        TestCase(String input, boolean expected) {
            this.input = input;
            this.expected = expected;
        }
    }
}

File: src/main/java/com/vahoss/kotlin_solutions/Solution00125.kt
package com.vahoss.kotlin_solutions

class Solution00125 {

    /**
     * O(n) time complexity using two pointer solution.
     * it's O(n) extra space!
     */
    fun isPalindrome(s: String): Boolean {
        val newS = s.lowercase().filter { it.isLetterOrDigit() }
        var beg = 0
        var end = newS.lastIndex

        while (end > beg) {
            val begCh = newS[beg]
            val endCh = newS[end]
            if (begCh != endCh) return false
            beg++
            end--
        }

        return true
    }

    fun isPalindromeOptimized(s: String): Boolean {
        var start = 0
        var end = s.lastIndex

        while (end > start) {
            while (!s[start].isLetterOrDigit() && end > start) start++
            val startCh = s[start].lowercaseChar()

            while (!s[end].isLetterOrDigit() && end > start) end--
            val endCh = s[end].lowercaseChar()

            if (startCh != endCh) return false
            start++
            end--
        }

        return true
    }
}

fun main() {
    val sol = Solution00125()
    val testCases = mapOf(
        "A man, a plan, a canal: Panama" to true,
        "race a car" to false,
        " " to true
    )

    testCases.forEach {
        val result = sol.isPalindrome(it.key)
        println("result is $result, expected is ${it.value}")
    }
}